### PR TITLE
Introduce opcode->identifier mapping X-Macro

### PR DESCRIPTION
**Pull request by chfast**
Use [X-Macro](https://en.wikipedia.org/wiki/X_Macro) (here having a pair of values) where a list of all opcodes is needed.

### CODE REVIEW

File: lib/evmone/advanced_instructions.cpp [hunk 1]
```diff
@@ -161,66 +170,12 @@ const Instruction* opx_beginblock(const Instruction* instr, AdvancedExecutionSta
...
-    table[OP_MLOAD] = op<mload>;
-    table[OP_MSTORE] = op<mstore>;
-    table[OP_MSTORE8] = op<mstore8>;
-    table[OP_SLOAD] = op<sload>;
+    // Init table with wrapped generic implementations.
+#define X(OPCODE, IDENTIFIER) table[OPCODE] = op<IDENTIFIER>;
+    MAP_OPCODE_TO_IDENTIFIER
+#undef X
+
+    // Overwrite with Advanced-specific implementations.
```
Comment (yperbasis) [inline lib/evmone/advanced_instructions.cpp:178]: `table[OP_MSIZE] = op<msize>` can be removed as well

File: lib/evmone/advanced_instructions.cpp [hunk 2]
```diff
@@ -161,66 +170,12 @@ const Instruction* opx_beginblock(const Instruction* instr, AdvancedExecutionSta
...
     table[OP_SSTORE] = op_sstore;
```
Comment (yperbasis) [inline lib/evmone/advanced_instructions.cpp:179]: Why is sstore overridden for the advanced interpreter but not sload?

Comment (chfast) [reply lib/evmone/advanced_instructions.cpp:179]: All instructions that need access to gas counter are replaced. This information is not directly available in the Advanced so we need to produce it for them. In particular `SSTORE` needs it if `gas < stipend`.

File: lib/evmone/advanced_instructions.cpp (hunk 1 above)
Comment (chfast) [reply lib/evmone/advanced_instructions.cpp:178]: Yes, done.

**Review by yperbasis — APPROVED** [review]

File: lib/evmone/advanced_instructions.cpp (hunk 2 above)
Comment (chfast) [reply lib/evmone/advanced_instructions.cpp:179]: No. I.e. many instruction are doing additional gas checks, but they don't  need to know exact value.

File: lib/evmone/advanced_instructions.cpp [hunk 3]
```diff
@@ -8,8 +8,16 @@
 
 namespace evmone::advanced
 {
+/// Fake wrap for generic instruction implementations accessing current code location.
+/// This is to make any op<...> compile, but pointers must be replaced with Advanced-specific
+/// implementation. Definition not provided.
+template <code_iterator InstrFn(ExecutionState&, code_iterator)>
+const Instruction* op(const Instruction* /*instr*/, AdvancedExecutionState& state) noexcept;
```
Comment (gumb0) [inline lib/evmone/advanced_instructions.cpp:15]: Can this be `= 0` maybe?

**Review by gumb0 — APPROVED** [review]
Well some of the C++ feel is lost to this :cry: 

But it's nice that in advanced it's now more clearly visible which implementations are advanced-specific.

File: lib/evmone/advanced_instructions.cpp (hunk 3 above)
Comment (chfast) [reply lib/evmone/advanced_instructions.cpp:15]: Do you mean `= delete`? I don't think so. The trick here is that this is actually used during compilation but later overwritten by Advanced's variants so linker does not complain. So this has to be external function definition (I even needed to move it out of anonymous namespace).

**Comment by chfast** [comment]
> Well some of the C++ feel is lost to this

This is really helpful in prototyping and also removes long blocks of 100+ lines of uninteresting code. This can be removed in future if instruction implementation signature stabilizes.

**Pull request merged**